src/app/admin/collectionCentresTable/CollectionCentresTable.tsx

```diff
@@ -1,116 +1,324 @@
-    const [collectionCentres, setCollectionCentres] = useState<Schema["collection_centres"][]>([]);
-    const [primaryFilters, setPrimaryFilters] =
-        useState<Filter<CollectionCentresTableRow, string>[]>(filters);
-    const [errorMessage, setErrorMessage] = useState<string | null>(null);
+interface EditToolbarProps {
+    setRows: (newRows: (oldRows: GridRowsProp) => GridRowsProp) => void;
+    setRowModesModel: (newModel: (oldModel: GridRowModesModel) => GridRowModesModel) => void;
+    rows: CollectionCentresTableRow[];
+}
+
+function EditToolbar(props: EditToolbarProps): React.JSX.Element {
+    const { setRows, setRowModesModel, rows } = props;
+
+    const handleClick = (): void => {
+        const id = rows.length + 1;
+        setRows((oldRows) => [...oldRows, { id, name: "", isShown: false, isNew: true }]);
+        setRowModesModel((oldModel) => ({
+            ...oldModel,
+            [id]: { mode: GridRowModes.Edit, fieldToFocus: "name" },
+        }));
+    };
+
+    return (
+        <GridToolbarContainer>
+            <Button color="primary" startIcon={<AddIcon />} onClick={handleClick}>
+                Add a new collection centre
+            </Button>
+        </GridToolbarContainer>
+    );
+}
 
-    const fetchAndDisplayCollectionCentres = useCallback(async () => {
-        const { data, error } = await supabase.from("collection_centres").select();
+function getBaseAuditLogForCollectionCentreAction(
+    action: string,
+    collectionCentreRow: CollectionCentresTableRow,
+    options?: {
+        excludeCollectionCentreId?: boolean;
+    }
+): Pick<AuditLog, "action" | "content" | "collectionCentreId"> {
+    return {
+        action,
+        content: {
+            collectionCentreName: collectionCentreRow.name,
+            collectionCentreAcronym: collectionCentreRow.acronym,
+            collectionCentreIsShown: collectionCentreRow.isShown,
+        },
+        collectionCentreId: options?.excludeCollectionCentreId ? undefined : collectionCentreRow.id,
+    };
+}
+
+const CollectionCentresTable: React.FC = () => {
+    const [rows, setRows] = useState<CollectionCentresTableRow[]>([]);
+    const [rowModesModel, setRowModesModel] = useState<GridRowModesModel>({});
+    const [isLoading, setIsLoading] = useState<boolean>(true);
+    const [errorMessage, setErrorMessage] = useState<string | null>(null);
 
+    const getCollectionCentresForTable = useCallback(async () => {
+        setErrorMessage(null);
+        const { data, error } = await fetchCollectionCentres();
         if (error) {
-            const logId = await logErrorReturnLogId("Error with fetch: Collection Centres", {
-                error: error,
-            });
-            setErrorMessage(`Failed to fetch collection centres. Log ID: ${logId}`);
+            setErrorMessage("Error fetching data, please reload");
             return;
         }
-
-        setCollectionCentres(data);
+        setRows(data);
+        setIsLoading(false);
```

Comment:
should we add a corresponding `setIsLoading(true)` before the fetch?

Reply:
The initial state of isLoading is true so I feel that it's not neccessary?

Reply:
ah fair enough!